### PR TITLE
Make InternalCentroid leaner

### PR DESCRIPTION
**Pull request by iverase**
We are currently holding to fields to extract values. Let's make them abstract methods so we don't use any heap. We can create several thousand of those objects on heap during reduction.

### CODE REVIEW

**Comment by iverase** [comment]
@elasticmachine update branch

File: server/src/main/java/org/elasticsearch/search/aggregations/metrics/InternalGeoCentroid.java [hunk 1]
```diff
@@ -84,12 +76,27 @@ protected InternalGeoCentroid copyWith(double firstSum, double secondSum, long t
...
+    @Override
+    protected double extractSecond(SpatialPoint point) {
+        return point.getX();
     }
 
-    static class Fields {
-        static final ParseField CENTROID_LAT = new ParseField("lat");
-        static final ParseField CENTROID_LON = new ParseField("lon");
+    @Override
+    public InternalAggregation finalizeSampling(SamplingContext samplingContext) {
```
Comment (craigtaverner) [inline server/src/main/java/org/elasticsearch/search/aggregations/metrics/InternalGeoCentroid.java:99]: This method can be deleted, or alternatively delete the copyWith method. Seems I added duplicates unnecessarily here.

**Review by craigtaverner — APPROVED** [review]
LGTM. I agree this generates less objects. I did notice one additional bit of code duplicate (from my earlier work) that could also be deleted.

**Comment by iverase** [comment]
@elasticmachine update branch

**Comment by iverase** [comment]
@elasticmachine update branch

**Comment by iverase** [comment]
@elasticmachine update branch

**Pull request merged**